webapp/src/snippetBuilder.tsx

```diff
@@ -0,0 +1,349 @@
+
+        return (
+            <sui.Modal isOpen={visible} className="snippetBuilder" size="large"
+                closeOnEscape={false} closeIcon={false} closeOnDimmerClick={false} closeOnDocumentClick={false}
+                dimmer={true} buttons={actions} header={config.name}
+            >
+                <div>
+                    <div className="list">
+                        {currQ &&
+                            <div>
+                                <div>{currQ.title}</div>
+                                <div className='list horizontal'>
+                                    {currQ.inputs.map((input: pxt.SnippetQuestionInput) =>
+                                        <div key={input.answerToken}>
+                                            <sui.Input
+                                                label={input.label && input.label}
+                                                onChange={this.textInputOnChange(input.answerToken)}
+                                                value={answers[input.answerToken] || ''}
+                                            />
+                                        </div>
+                                    )}
+                                </div>
+                            </div>
+                        }
+                    </div>
+                    <div className='snippetBuilderOutput'>
+                        {parent && <md.MarkedContent markdown={this.generateOutputMarkdown(tsOutput)} parent={parent} />}
+                    </div>
+                </div>
+            </sui.Modal>
+        )
+    }
+}
+
+function getSnippetExtensions() {
+    return pkg
+        .allEditorPkgs()
+        .map(ep => ep.getKsPkg())
+        .map(p => !!p && p.config)
+        .filter(config => config.snippetBuilders);
+}
+
+function openSnippetDialog(config: pxt.SnippetConfig, editor: Blockly.WorkspaceSvg, parent: pxt.editor.IProjectView) {
+    const wrapper = document.body.appendChild(document.createElement('div'));
+    const props = { parent: parent, mainWorkspace: editor, config };
+    const snippetBuilder = ReactDOM.render(
+        React.createElement(SnippetBuilder, props),
+        wrapper
+    ) as SnippetBuilder;
+    snippetBuilder.show();
+}
+
+export function initializeSnippetExtensions(ns: string, extraBlocks: (toolbox.BlockDefinition | toolbox.ButtonDefinition)[], editor: Blockly.WorkspaceSvg, parent: pxt.editor.IProjectView) {
+    const snippetExtensions = getSnippetExtensions();
+
+    snippetExtensions.forEach(config => {
+        config.snippetBuilders.forEach(snippet => {
+            if (ns == snippet.namespace) {
+                extraBlocks.push({
+                    name: `SNIPPET${name}_BUTTON`,
+                    type: "button",
+                    attributes: {
+                        blockId: `SNIPPET${name}_BUTTON`,
+                        label: snippet.label ? pxt.Util.rlf(snippet.label) : pxt.Util.lf("Editor"),
```

Comment:
strings need to be injected in target-strings.json when the CLI builds the target